File: src/net/java/sip/communicator/plugin/addrbook/msoutlook/MsOutlookMAPIHResultException.java
/*
 * Jitsi, the OpenSource Java VoIP and Instant Messaging client.
 *
 * Copyright @ 2015 Atlassian Pty Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.java.sip.communicator.plugin.addrbook.msoutlook;

/**
 * Represents a specific Microsoft Outlook MAPI <tt>HRESULT</tt> as an
 * <tt>Exception</tt>.
 *
 * @author Lyubomir Marinov
 */
public class MsOutlookMAPIHResultException
        extends Exception
{
    /**
     * Serial version UID.
     */
    private static final long serialVersionUID = 0L;

    /**
     * The <tt>HRESULT</tt> which is represented by this <tt>Exception</tt>.
     */
    private final long hResult;

    /**
     * Initializes a new <tt>MsOutlookMAPIHResultException</tt> instance which
     * is to represent a specific <tt>HRESULT</tt>.
     *
     * @param hResult the <tt>HRESULT</tt> to be represented by the new instance
     */
    public MsOutlookMAPIHResultException(long hResult)
    {
        this(hResult, toString(hResult));
    }

    /**
     * Initializes a new <tt>MsOutlookMAPIHResultException</tt> instance which
     * is to represent a specific <tt>HRESULT</tt> and to provide a specific
     * <tt>String</tt> message.
     *
     * @param hResult the <tt>HRESULT</tt> to be represented by the new instance
     * @param message the <tt>String</tt> message to be provided by the new
     * instance
     */
    public MsOutlookMAPIHResultException(long hResult, String message)
    {
        super(message);

        this.hResult = hResult;
    }

    /**
     * Initializes a new <tt>MsOutlookMAPIHResultException</tt> instance with a
     * specific <tt>String</tt> message.
     *
     * @param message the <tt>String</tt> message to be provided by the new
     * instance
     */
    public MsOutlookMAPIHResultException(String message)
    {
        this(0, message);
    }

    /**
     * Gets the <tt>HRESULT</tt> which is represented by this
     * <tt>Exception</tt>.
     *
     * @return the <tt>HRESULT</tt> which is represented by this
     * <tt>Exception</tt>
     */
    public long getHResult()
    {
        return hResult;
    }

    /**
     * Converts a specific <tt>HRESULT</tt> to a touch more readable
     * <tt>String</tt> in accord with the rule of constructing MAPI
     * <tt>HRESULT</tt> values.
     *
     * @param hResult the <tt>HRESULT</tt> to convert
     * @return a <tt>String</tt> which represents the specified <tt>hResult</tt>
     * in a touch more readable form
     */
    private static String toString(long hResult)
    {
        if (hResult == 0)
            return "S_OK";
        else
        {
            StringBuilder s = new StringBuilder("MAPI_");

            s.append(((hResult & 0x80000000L) == 0) ? 'W' : 'E');
            s.append("_0x");
            s.append(Long.toHexString(hResult & 0xFFFL));
            s.append(" (" + Long.toHexString(hResult) + ")");
            return s.toString();
        }
    }

    /**
     * Returns the string representation for the current hResult code.
     *
     * @return The string representation for the current hResult code.
     */
    public String getHresultString()
    {
        return MsOutlookMAPIHResultException.toString(this.getHResult());
    }
}

File: src/native/addrbook/msoutlook/MsOutlookMAPIHResultException.cxx
/*
 * Jitsi, the OpenSource Java VoIP and Instant Messaging client.
 *
 * Copyright @ 2015 Atlassian Pty Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "MsOutlookMAPIHResultException.h"

void
MsOutlookMAPIHResultException_throwNew
    (JNIEnv *jniEnv, HRESULT hResult, LPCSTR file, ULONG line)
{
    jclass clazz;

    clazz
        = jniEnv->FindClass(
                "net/java/sip/communicator/plugin/addrbook/msoutlook/MsOutlookMAPIHResultException");
    if (clazz)
    {
        LPCSTR message;

        switch (hResult)
        {
        case MAPI_E_LOGON_FAILED:
            message = "MAPI_E_LOGON_FAILED";
            break;
        case MAPI_E_NO_ACCESS:
            message = "MAPI_E_NO_ACCESS";
            break;
        case MAPI_E_NO_SUPPORT:
            message = "MAPI_E_NO_SUPPORT";
            break;
        case MAPI_E_NOT_ENOUGH_MEMORY:
            message = "MAPI_E_NOT_ENOUGH_MEMORY";
            break;
        case MAPI_E_NOT_FOUND:
            message = "MAPI_E_NOT_FOUND";
            break;
        case MAPI_E_NOT_INITIALIZED:
            message = "MAPI_E_NOT_INITIALIZED";
            break;
        case MAPI_E_TIMEOUT:
            message = "MAPI_E_TIMEOUT";
            break;
        case MAPI_E_UNKNOWN_ENTRYID:
            message = "MAPI_E_UNKNOWN_ENTRYID";
            break;
        case MAPI_E_USER_CANCEL:
            message = "MAPI_E_USER_CANCEL";
            break;
        case MAPI_W_ERRORS_RETURNED:
            message = "MAPI_W_ERRORS_RETURNED";
            break;
        case S_OK:
            message = "S_OK";
            break;
        default:
            message = NULL;
            break;
        }

        if (message)
        {
            jmethodID methodID
                = jniEnv->GetMethodID(
                        clazz,
                        "<init>",
                        "(JLjava/lang/String;)V");

            if (methodID)
            {
                jstring jmessage = jniEnv->NewStringUTF(message);

                if (jmessage)
                {
                    jobject t
                        = jniEnv->NewObject(
                                clazz,
                                methodID,
                                (jlong) hResult, jmessage);

                    if (t)
                    {
                        jniEnv->Throw((jthrowable) t);

                        jniEnv->DeleteLocalRef(t);
                    }
                    jniEnv->DeleteLocalRef(jmessage);
                }
                return;
            }
        }

        {
            jmethodID methodID = jniEnv->GetMethodID(clazz, "<init>", "(J)V");

            if (methodID)
            {
                jobject t = jniEnv->NewObject(clazz, methodID, hResult);

                if (t)
                {
                    jniEnv->Throw((jthrowable) t);

                    jniEnv->DeleteLocalRef(t);
                }
                return;
            }
        }

        jniEnv->ThrowNew(clazz, message);

        jniEnv->DeleteLocalRef(clazz);
    }
}
